### PR TITLE
Tensorflow Script for DCGAN

### PR DESCRIPTION
**Pull request by merveenoyan**
This PR has tensorflow training script for DCGAN. I feel like we can just adapt it for other architectures as well.

### CODE REVIEW

File: huggan/tensorflow/dcgan/train.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        img = np.asarray(img) 
+        img = tf.keras.utils.img_to_array(PIL.Image.fromarray(img))
+        preprocessed_images.append(img)
+    train_images = np.asarray(preprocessed_images)
+    channel = train_images[0].shape[-1]
+    train_images = train_images.reshape(train_images.shape[0], image_size, image_size, channel).astype('float32')
+
+    train_images = (train_images - 127.5) / 127.5 
+    train_dataset = tf.data.Dataset.from_tensor_slices(train_images).batch(BATCH_SIZE)
+    return train_dataset, channel
```
Comment (NielsRogge) [inline huggan/tensorflow/dcgan/train.py:191]: Hmm not sure if it should be processed like that. The big features of Datasets are .map, .with_transform and .set_transform (and optionally set_format).

You shouldn't manually iterate over the rows of the dataset, the aformentioned functions do that for you (see also my Pytorch script: the function you define should just be defined for a batch of examples)

Comment (merveenoyan) [reply huggan/tensorflow/dcgan/train.py:191]: I had a problem in process and wanted to get script out ASAP. I can do this if you want that.

Comment (merveenoyan) [reply huggan/tensorflow/dcgan/train.py]: I did this instead, `PrefetchDataset` returned by `to_tf_dataset` doesn't work well with `model()` and dataset map was returning elements like [[image],[another image]] so I unpacked and made another list that `from_tensor_slices` wanted.
```
def preprocess_images(dataset, args):
    processed_dataset = dataset["test"].map(preprocess)
    new_dataset = []
    
    for image in processed_dataset["pixel_values"]:
        new_dataset.append(image[0])
    
    new_dataset = tf.data.Dataset.from_tensor_slices(new_dataset).batch(args.batch_size)
    return new_dataset
```

Comment (NielsRogge) [reply huggan/tensorflow/dcgan/train.py]: cc'ing @Rocketknight1 here, apparently` .to_tf_dataset` wasn't ideal for vision datasets

Comment (merveenoyan) [reply huggan/tensorflow/dcgan/train.py]: added `to_tf_dataset` as well

Comment (merveenoyan) [reply huggan/tensorflow/dcgan/train.py]: `def preprocess(examples):
    images = (np.asarray(examples["image"]).astype('float32')- 127.5) / 127.5
    images = np.expand_dims(images, -1)
    examples["pixel_values"] = images
    return examples



def preprocess_images(dataset, args):
    
 
    data_collator = DefaultDataCollator(return_tensors="tf")


    processed_dataset = dataset.map(preprocess)
    
    tf_train_dataset = processed_dataset.to_tf_dataset(
	    columns=['pixel_values'],
	    shuffle=True,
	    batch_size=args.batch_size,
	    collate_fn=data_collator)

    return tf_train_dataset`

I didn't expand dimensions previously that's why it wasn't working

File: huggan/tensorflow/dcgan/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+## Train DCGAN on your custom data
+This folder contains a script to train DCGAN for unconditional image generation, leveraging the Hugging Face ecosystem for processing your data and pushing the model to the Hub.
+
+The script leverages 🤗 Datasets for loading and processing data, and TensorFlow for training the model and 🤗 Hub for hosting it.
```
Comment (NielsRogge) [inline huggan/tensorflow/dcgan/README.md:4]:
```suggestion
The script leverages 🤗 Datasets for loading and processing data, TensorFlow for training the model and 🤗 Hub for hosting it.
```

File: huggan/tensorflow/dcgan/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+## Launching the script
+You can simply run `python train.py` with default parameters. It will download MNIST dataset, preprocess it and train a model on it, will save results of epochs in a local directory and push to 🤗 Hub.
```
Comment (NielsRogge) [inline huggan/tensorflow/dcgan/README.md:7]:
```suggestion
You can simply run `python train.py` with the default parameters. It will download the MNIST dataset, preprocess it and train a model on it, will save results after each epoch in a local directory and push the model to the 🤗 Hub.
```

File: huggan/tensorflow/dcgan/README.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+To train on another dataset available on the hub, simply do (for instance):
+
+python train.py --dataset cifar-10
+In case you'd like to tweak the script to your liking, first fork the "community-events" repo (see the button on the top right), then clone it locally:
```
Comment (NielsRogge) [inline huggan/tensorflow/dcgan/README.md:12]:
```suggestion
```bash
python train.py --dataset cifar-10
```

In case you'd like to tweak the script to your liking, first fork the "community-events" repo (see the button on the top right), then clone it locally:
```

File: huggan/tensorflow/dcgan/README.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+# next: push to the hub (assuming git-LFS is installed)
+dataset.push_to_hub("huggan/my-awesome-dataset")
+# You can then simply pass the name of the dataset to the script:
+
+python train.py --dataset huggan/my-awesome-dataset
+```
+## Pushing model to the Hub
+
+For this you can use `push_to_hub_keras` which generates a card for your model with training metrics, plot of the architecture and hyperparameters. For this, specify --output_dir and --model_name and use --push_to_hub flag like so:
```
Comment (NielsRogge) [inline huggan/tensorflow/dcgan/README.md:45]:
```suggestion
For this you can use `push_to_hub_keras` which generates a card for your model with training metrics, plot of the architecture and hyperparameters. For this, specify `--output_dir` and `--model_name` and use the `--push_to_hub` flag like so:
```

File: huggan/tensorflow/dcgan/train.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+
+
+def preprocess_images(dataset, args):
+    
+ 
+    data_collator = DefaultDataCollator(return_tensors="tf")
+
+
+    processed_dataset = dataset.map(preprocess)
```
Comment (NielsRogge) [inline huggan/tensorflow/dcgan/train.py:196]:
```suggestion
def preprocess_images(dataset, args):
    data_collator = DefaultDataCollator(return_tensors="tf")
    processed_dataset = dataset.map(preprocess)
```

File: huggan/tensorflow/dcgan/README.md (hunk 2 above)
Comment (osanseviero) [reply huggan/tensorflow/dcgan/README.md:4]: Actually grammar-wise the current one is already correct I think :smile:

Comment (NielsRogge) [reply huggan/tensorflow/dcgan/README.md:4]: Haha yeah sorry this is a major nitpick. I usually only include "and" at the end

**Comment by osanseviero** [comment]
Maybe let's also add it to https://github.com/huggingface/community-events/tree/main/huggan#example-scripts

**Comment by NielsRogge** [comment]
Ok I'll merge it and add a link
